File: electron/ipc/window.ts
import { ipcMain, BrowserWindow } from 'electron';

type Result<T> = { success: true; data: T } | { success: false; error: string };

export function registerWindowHandlers(window: BrowserWindow): void {
  ipcMain.handle('window:minimize', async (): Promise<Result<null>> => {
    try {
      window.minimize();
      return { success: true, data: null };
    } catch (err) {
      return { success: false, error: (err as Error).message };
    }
  });

  ipcMain.handle('window:maximize', async (): Promise<Result<null>> => {
    try {
      if (window.isMaximized()) {
        window.unmaximize();
      } else {
        window.maximize();
      }
      return { success: true, data: null };
    } catch (err) {
      return { success: false, error: (err as Error).message };
    }
  });

  ipcMain.handle('window:close', async (): Promise<Result<null>> => {
    try {
      window.close();
      return { success: true, data: null };
    } catch (err) {
      return { success: false, error: (err as Error).message };
    }
  });
}

File: src-tauri/src/commands/window.rs
use tauri::{AppHandle, Manager};

#[derive(serde::Serialize)]
pub struct WindowResult<T> {
    pub success: bool,
    pub data: Option<T>,
    pub error: Option<String>,
}

#[tauri::command]
pub async fn window_minimize(app: AppHandle) -> WindowResult<()> {
    if let Some(window) = app.get_webview_window("main") { let _ = window.minimize(); }
    WindowResult { success: true, data: Some(()), error: None }
}

#[tauri::command]
pub async fn window_maximize(app: AppHandle) -> WindowResult<()> {
    if let Some(window) = app.get_webview_window("main") {
        if window.is_maximized().unwrap_or(false) { let _ = window.unmaximize(); }
        else { let _ = window.maximize(); }
    }
    WindowResult { success: true, data: Some(()), error: None }
}

#[tauri::command]
pub async fn window_close(app: AppHandle) -> WindowResult<()> {
    if let Some(window) = app.get_webview_window("main") { let _ = window.close(); }
    WindowResult { success: true, data: Some(()), error: None }
}

#[tauri::command]
pub async fn window_show(app: AppHandle) -> WindowResult<()> {
    if let Some(window) = app.get_webview_window("main") {
        let _ = window.show();
        let _ = window.set_focus();
    }
    WindowResult { success: true, data: Some(()), error: None }
}

#[tauri::command]
pub async fn window_hide(app: AppHandle) -> WindowResult<()> {
    if let Some(window) = app.get_webview_window("main") { let _ = window.hide(); }
    WindowResult { success: true, data: Some(()), error: None }
}
